### PR TITLE
Use std::vector for tags, members and nodes

### PR DESCRIPTION
**Pull request by lonvia**
Replaces keyval and other manually malloc'd lists with vectors. Make output data from parser const and simply create new vectors for taglists when they need to be changed.

Also optionally enables C++11 in configure where available but still without using any C++11 syntax in the code. This gives us move semantics for vectors where available which improves performance slightly.

### CODE REVIEW

**Comment by pnorman** [comment]
travis is failing with 
```
./configure: line 5194: syntax error near unexpected token `,optional'
./configure: line 5194: `AX_CXX_COMPILE_STDCXX_11(,optional)'
```

**Comment by pnorman** [comment]
Adding the file and building with clang++ 3.4 I get
```
depbase=`echo osmdata.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
/bin/sh ./libtool  --tag=CXX   --mode=compile c++ -DHAVE_CONFIG_H -I.  -D_THREAD_SAFE -pthread -I/usr/local/include -I/usr/local/include/libxml2 -I/usr/include -I/usr/include -I/usr/local/include -I/usr/local/include -DOSM2PGSQL_DATADIR='"/usr/local/share/osm2pgsql"' -Igeos-fallback -I/usr/local/include/lua52 -I/usr/local/include   -g -O2 -std=gnu++11 -MT osmdata.lo -MD -MP -MF $depbase.Tpo -c -o osmdata.lo osmdata.cpp &&\
mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  c++ -DHAVE_CONFIG_H -I. -D_THREAD_SAFE -pthread -I/usr/local/include -I/usr/local/include/libxml2 -I/usr/include -I/usr/include -I/usr/local/include -I/usr/local/include -DOSM2PGSQL_DATADIR=\"/usr/local/share/osm2pgsql\" -Igeos-fallback -I/usr/local/include/lua52 -I/usr/local/include -g -O2 -std=gnu++11 -MT osmdata.lo -MD -MP -MF .deps/osmdata.Tpo -c osmdata.cpp  -fPIC -DPIC -o .libs/osmdata.o
In file included from osmdata.cpp:1:
In file included from ./osmdata.hpp:8:
In file included from /usr/local/include/boost/shared_ptr.hpp:17:
In file included from /usr/local/include/boost/smart_ptr/shared_ptr.hpp:27:
In file included from /usr/local/include/boost/config/no_tr1/memory.hpp:21:
In file included from /usr/include/c++/v1/memory:603:
/usr/include/c++/v1/tuple:320:11: error: rvalue reference to type 'boost::thread_detail::default_barrier_reseter' cannot bind to lvalue of type 'boost::thread_detail::default_barrier_reseter'
        : value(__t.get())
          ^     ~~~~~~~~~
/usr/include/c++/v1/tuple:444:8: note: in instantiation of member function 'std::__1::__tuple_leaf<0, boost::thread_detail::default_barrier_reseter &&, false>::__tuple_leaf' requested here
struct __tuple_impl<__tuple_indices<_Indx...>, _Tp...>
       ^
/usr/include/c++/v1/functional:1286:26: note: in instantiation of member function 'std::__1::__function::__func<boost::thread_detail::default_barrier_reseter, std::__1::allocator<boost::thread_detail::default_barrier_reseter>, unsigned long ()>::__func' requested here
            ::new (__f_) _FF(_VSTD::move(__f));
                         ^
/usr/local/include/boost/thread/barrier.hpp:108:55: note: in instantiation of function template specialization 'std::__1::function<unsigned long ()>::function<boost::thread_detail::default_barrier_reseter>' requested here
      m_count(check_counter(count)), m_generation(0), fct_(thread_detail::default_barrier_reseter(count))
                                                      ^
In file included from osmdata.cpp:1:
In file included from ./osmdata.hpp:8:
In file included from /usr/local/include/boost/shared_ptr.hpp:17:
In file included from /usr/local/include/boost/smart_ptr/shared_ptr.hpp:27:
In file included from /usr/local/include/boost/config/no_tr1/memory.hpp:21:
In file included from /usr/include/c++/v1/memory:603:
/usr/include/c++/v1/tuple:321:10: error: static_assert failed "Can not copy a tuple with rvalue reference member"
        {static_assert(!is_rvalue_reference<_Hp>::value, "Can not copy a tuple with rvalue reference member");}
         ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/tuple:320:11: error: rvalue reference to type 'allocator<[...]>' cannot bind to lvalue of type 'allocator<[...]>'
        : value(__t.get())
          ^     ~~~~~~~~~
/usr/include/c++/v1/tuple:444:8: note: in instantiation of member function 'std::__1::__tuple_leaf<0, std::__1::allocator<boost::thread_detail::default_barrier_reseter> &&, false>::__tuple_leaf' requested here
struct __tuple_impl<__tuple_indices<_Indx...>, _Tp...>
       ^
/usr/include/c++/v1/functional:1294:34: note: in instantiation of member function 'std::__1::__function::__func<boost::thread_detail::default_barrier_reseter, std::__1::allocator<boost::thread_detail::default_barrier_reseter>, unsigned long ()>::__func' requested here
            ::new (__hold.get()) _FF(_VSTD::move(__f), allocator<_Fp>(__a));
                                 ^
/usr/local/include/boost/thread/barrier.hpp:108:55: note: in instantiation of function template specialization 'std::__1::function<unsigned long ()>::function<boost::thread_detail::default_barrier_reseter>' requested here
      m_count(check_counter(count)), m_generation(0), fct_(thread_detail::default_barrier_reseter(count))
                                                      ^
In file included from osmdata.cpp:1:
In file included from ./osmdata.hpp:8:
In file included from /usr/local/include/boost/shared_ptr.hpp:17:
In file included from /usr/local/include/boost/smart_ptr/shared_ptr.hpp:27:
In file included from /usr/local/include/boost/config/no_tr1/memory.hpp:21:
In file included from /usr/include/c++/v1/memory:603:
/usr/include/c++/v1/tuple:321:10: error: static_assert failed "Can not copy a tuple with rvalue reference member"
        {static_assert(!is_rvalue_reference<_Hp>::value, "Can not copy a tuple with rvalue reference member");}
         ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/tuple:320:11: error: rvalue reference to type 'boost::thread_detail::void_fct_ptr_barrier_reseter' cannot bind to lvalue of type 'boost::thread_detail::void_fct_ptr_barrier_reseter'
        : value(__t.get())
          ^     ~~~~~~~~~
/usr/include/c++/v1/tuple:444:8: note: in instantiation of member function 'std::__1::__tuple_leaf<0, boost::thread_detail::void_fct_ptr_barrier_reseter &&, false>::__tuple_leaf' requested here
struct __tuple_impl<__tuple_indices<_Indx...>, _Tp...>
       ^
/usr/include/c++/v1/functional:1286:26: note: in instantiation of member function 'std::__1::__function::__func<boost::thread_detail::void_fct_ptr_barrier_reseter, std::__1::allocator<boost::thread_detail::void_fct_ptr_barrier_reseter>, unsigned long ()>::__func' requested here
            ::new (__f_) _FF(_VSTD::move(__f));
                         ^
/usr/local/include/boost/thread/barrier.hpp:164:13: note: in instantiation of function template specialization 'std::__1::function<unsigned long ()>::function<boost::thread_detail::void_fct_ptr_barrier_reseter>' requested here
          ? thread_detail::size_completion_function(thread_detail::void_fct_ptr_barrier_reseter(count, funct))
            ^
In file included from osmdata.cpp:1:
In file included from ./osmdata.hpp:8:
In file included from /usr/local/include/boost/shared_ptr.hpp:17:
In file included from /usr/local/include/boost/smart_ptr/shared_ptr.hpp:27:
In file included from /usr/local/include/boost/config/no_tr1/memory.hpp:21:
In file included from /usr/include/c++/v1/memory:603:
/usr/include/c++/v1/tuple:321:10: error: static_assert failed "Can not copy a tuple with rvalue reference member"
        {static_assert(!is_rvalue_reference<_Hp>::value, "Can not copy a tuple with rvalue reference member");}
         ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/tuple:320:11: error: rvalue reference to type 'allocator<[...]>' cannot bind to lvalue of type 'allocator<[...]>'
        : value(__t.get())
          ^     ~~~~~~~~~
/usr/include/c++/v1/tuple:444:8: note: in instantiation of member function 'std::__1::__tuple_leaf<0, std::__1::allocator<boost::thread_detail::void_fct_ptr_barrier_reseter> &&, false>::__tuple_leaf' requested here
struct __tuple_impl<__tuple_indices<_Indx...>, _Tp...>
       ^
/usr/include/c++/v1/functional:1294:34: note: in instantiation of member function 'std::__1::__function::__func<boost::thread_detail::void_fct_ptr_barrier_reseter, std::__1::allocator<boost::thread_detail::void_fct_ptr_barrier_reseter>, unsigned long ()>::__func' requested here
            ::new (__hold.get()) _FF(_VSTD::move(__f), allocator<_Fp>(__a));
                                 ^
/usr/local/include/boost/thread/barrier.hpp:164:13: note: in instantiation of function template specialization 'std::__1::function<unsigned long ()>::function<boost::thread_detail::void_fct_ptr_barrier_reseter>' requested here
          ? thread_detail::size_completion_function(thread_detail::void_fct_ptr_barrier_reseter(count, funct))
            ^
In file included from osmdata.cpp:1:
In file included from ./osmdata.hpp:8:
In file included from /usr/local/include/boost/shared_ptr.hpp:17:
In file included from /usr/local/include/boost/smart_ptr/shared_ptr.hpp:27:
In file included from /usr/local/include/boost/config/no_tr1/memory.hpp:21:
In file included from /usr/include/c++/v1/memory:603:
/usr/include/c++/v1/tuple:321:10: error: static_assert failed "Can not copy a tuple with rvalue reference member"
        {static_assert(!is_rvalue_reference<_Hp>::value, "Can not copy a tuple with rvalue reference member");}
         ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/tuple:320:11: error: rvalue reference to type 'unsigned int (*)()' cannot bind to lvalue of type 'unsigned int (*)()'
        : value(__t.get())
          ^     ~~~~~~~~~
/usr/include/c++/v1/tuple:444:8: note: in instantiation of member function 'std::__1::__tuple_leaf<0, unsigned int (*&&)(), false>::__tuple_leaf' requested here
struct __tuple_impl<__tuple_indices<_Indx...>, _Tp...>
       ^
/usr/include/c++/v1/functional:1286:26: note: in instantiation of member function 'std::__1::__function::__func<unsigned int (*)(), std::__1::allocator<unsigned int (*)()>, unsigned long ()>::__func' requested here
            ::new (__f_) _FF(_VSTD::move(__f));
                         ^
/usr/local/include/boost/thread/barrier.hpp:172:13: note: in instantiation of function template specialization 'std::__1::function<unsigned long ()>::function<unsigned int (*)()>' requested here
          ? thread_detail::size_completion_function(funct)
            ^
In file included from osmdata.cpp:1:
In file included from ./osmdata.hpp:8:
In file included from /usr/local/include/boost/shared_ptr.hpp:17:
In file included from /usr/local/include/boost/smart_ptr/shared_ptr.hpp:27:
In file included from /usr/local/include/boost/config/no_tr1/memory.hpp:21:
In file included from /usr/include/c++/v1/memory:603:
/usr/include/c++/v1/tuple:321:10: error: static_assert failed "Can not copy a tuple with rvalue reference member"
        {static_assert(!is_rvalue_reference<_Hp>::value, "Can not copy a tuple with rvalue reference member");}
         ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/tuple:320:11: error: rvalue reference to type 'allocator<[...]>' cannot bind to lvalue of type 'allocator<[...]>'
        : value(__t.get())
          ^     ~~~~~~~~~
/usr/include/c++/v1/tuple:444:8: note: in instantiation of member function 'std::__1::__tuple_leaf<0, std::__1::allocator<unsigned int (*)()> &&, false>::__tuple_leaf' requested here
struct __tuple_impl<__tuple_indices<_Indx...>, _Tp...>
       ^
/usr/include/c++/v1/functional:1294:34: note: in instantiation of member function 'std::__1::__function::__func<unsigned int (*)(), std::__1::allocator<unsigned int (*)()>, unsigned long ()>::__func' requested here
            ::new (__hold.get()) _FF(_VSTD::move(__f), allocator<_Fp>(__a));
                                 ^
/usr/local/include/boost/thread/barrier.hpp:172:13: note: in instantiation of function template specialization 'std::__1::function<unsigned long ()>::function<unsigned int (*)()>' requested here
          ? thread_detail::size_completion_function(funct)
            ^
In file included from osmdata.cpp:1:
In file included from ./osmdata.hpp:8:
In file included from /usr/local/include/boost/shared_ptr.hpp:17:
In file included from /usr/local/include/boost/smart_ptr/shared_ptr.hpp:27:
In file included from /usr/local/include/boost/config/no_tr1/memory.hpp:21:
In file included from /usr/include/c++/v1/memory:603:
/usr/include/c++/v1/tuple:321:10: error: static_assert failed "Can not copy a tuple with rvalue reference member"
        {static_assert(!is_rvalue_reference<_Hp>::value, "Can not copy a tuple with rvalue reference member");}
         ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12 errors generated.
Makefile:1185: recipe for target 'osmdata.lo' failed
gmake: *** [osmdata.lo] Error 1
```

This is something with C++11, as 8d803dc4d72f155dd89968916da12de297280c7f doesn't get that error.

**Comment by lonvia** [comment]
I've reverted the C++11 change. Obviously needs some more work first. The rest is still ok as it is.

File: expire-tiles.cpp [hunk 1]
```diff
@@ -413,80 +413,71 @@ int expire_tiles::from_bbox(double min_lon, double min_lat, double max_lon, doub
...
-	}
+void expire_tiles::from_nodes_line(const nodelist_t &nodes)
+{
+    if (Options->expire_tiles_zoom < 0 || nodes.empty())
+        return;
+
+    if (nodes.size() == 1) {
+        from_bbox(nodes[0].lon, nodes[0].lat, nodes[0].lon, nodes[0].lat);
+    } else {
+        for (int i = 1; i < nodes.size(); ++i)
```
Comment (lonvia) [inline expire-tiles.cpp]: In theory yes, but I found that version much more readable. With nodes being const, I trust that the compiler knows how to optimize the code correctly.

Comment (pnorman) [inline expire-tiles.cpp]: k. i guess with C++11 later on we can move to auto eventually ;)

**Comment by lonvia** [comment]
Some more changes: there were a lot of new warnings about signed vs. unsigned comparison I missed because warnings were disabled. I've enabled them now and fixed all warnings. Also there was still an issue where the wrong way_area was written out because the reading order of the taglist has changed.

**Comment by lonvia** [comment]
@pnorman I've tried compiling with C++11 and clang and it runs through without any problems. Tested against clang 3.5 with libc++, boost 1.55 and clang 3.4 with libstdc++, boost 1.55. The error seems to come from somewhere inside boost or libc++. Which version are you using?

**Comment by pnorman** [comment]
> The error seems to come from somewhere inside boost or libc++. Which version are you using?

clang 3.4.1, boost 1.5.5, libc++. The error only showed on `make check TESTS=''`, not on `make`

It now builds successfully. No warnings on `make`, some on `make check TESTS=''`

**Comment by pnorman** [comment]
no change or slightly faster in the parsing stage. I have a rule about not merging this late at night so I'm not going to do it, but looks okay to me.

**Pull request merged**